google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java

```diff
@@ -0,0 +1,256 @@
+
+import org.jetbrains.annotations.Nullable;
+
+import java.awt.BorderLayout;
+import java.awt.Color;
+import java.awt.Desktop;
+import java.awt.Dimension;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Enumeration;
+
+import javax.swing.BorderFactory;
+import javax.swing.Box;
+import javax.swing.BoxLayout;
+import javax.swing.JButton;
+import javax.swing.JPanel;
+import javax.swing.JTree;
+import javax.swing.tree.DefaultMutableTreeNode;
+import javax.swing.tree.DefaultTreeModel;
+import javax.swing.tree.TreeNode;
+import javax.swing.tree.TreeSelectionModel;
+
+/**
+ * Defines a repository selector UI widget allowing the user to select a repository associated
+ * with a given Google Cloud project.
+ */
+public class RepositorySelector extends CustomizableComboBox implements CustomizableComboBoxPopup {
+
+  private static final Logger logger = Logger.getInstance(RepositorySelector.class);
+
+  private JBPopup popup;
+  private RepositoryPanel panel;
+  private String cloudProject;
+  private CredentialedUser user;
+  private boolean canCreateRepository;
+  private static final int SELECTOR_HEIGHT = 140;
+  private static final int SELECTOR_WIDTH = 400;
+
+
+  public RepositorySelector(@Nullable String cloudProject,
+      @Nullable CredentialedUser user, boolean canCreateRepository) {
+    this.cloudProject = cloudProject;
+    this.user = user;
+    this.canCreateRepository = canCreateRepository;
+
+    getTextField().getEmptyText().setText("Enter a cloud repository ID...");
+  }
+
+  @Nullable
+  public String getSelectedRepository() {
+    if (StringUtil.isEmpty(getText())) {
+      return null;
+    }
+
+    Enumeration repos = panel.getRepositories().children();
+    while (repos.hasMoreElements()) {
+      TreeNode repo = (TreeNode) repos.nextElement();
+
+      if (repo instanceof RepositoryModelItem
+          && getText().equals(((RepositoryModelItem) repo).getRepositoryId())) {
+        return getText();
+      }
+    }
+
+    return null;
+  }
+
+  public void setCloudProject(String cloudProject) {
+    this.cloudProject = cloudProject;
+  }
+
+  public void setUser(CredentialedUser user) {
+    this.user = user;
+  }
+
+  @Override
+  public void showPopup(RelativePoint showTarget) {
+    if (user != null) {
+      if (popup == null || popup.isDisposed()) {
+        panel = new RepositoryPanel(cloudProject);
+
+        ComponentPopupBuilder popupBuilder = JBPopupFactory.getInstance()
+            .createComponentPopupBuilder(panel, null);
+        popup = popupBuilder.createPopup();
+      }
+      if (!popup.isVisible()) {
+        popup.show(showTarget);
+      }
+    }
+  }
+
+  @Override
+  public void hidePopup() {
+    if (isPopupVisible()) {
+      popup.closeOk(null);
+    }
+  }
+
+  @Override
+  public boolean isPopupVisible() {
+    return popup != null && !popup.isDisposed() && popup.isVisible();
+  }
+
+  @Override
+  protected CustomizableComboBoxPopup getPopup() {
+    return this;
```

Comment:
I wonder what is the use of a method that returns the reference of the object it's being called on?